Cartesian: "No Storage" Implementation

gronerl opened the pull request:
This PR removes the Storage classes and implements the new "No Storage" concept.

It is currently fully in the cartesian branch but the same shall be applied to `functional`.

gronerl commented:
bors try

gronerl commented:
bors try-

gronerl commented:
bors try

gronerl commented:
bors try

gronerl commented:
bors try

gronerl commented:
bors try-
bors try
😐

gronerl commented:
bors try

gronerl commented:
bors try

gronerl commented:
bors try-
bors try

jdahm commented on docs/gt4py/quickstart.rst:
```diff
@@ -183,29 +185,43 @@ set of routines used in `NumPy` for allocation: ``from_array``, ``ones``, ``zero
 
 
-The ``default_origin`` parameter plays two roles:
+The ``aligned_index`` specifies that the array is to be allocated such that memory address of the point specified in
+``aligned_index`` is `aligned` to a backend-dependent value. For optimal performance, you set the ``algined_index`` to
+a point which is the lower-left corner of the iteration domain most frequently used for this field.
```
Still diving in here so apologies if this is clear later: does specifying `aligned_index` instead of `default_origin` mean that we always need to specify non-zero origins when calling stencils, or is this used as the default origin too?

gronerl replied:
Kind of. If you just rely on the allocation routines, you get an unmodified NumPy ndarray. That is, there is no `default_origin` attribute like there used to be on the storages. However, if a user passes a buffer object that (This could e.g. be a `Quantity` in pace) They can opt to implenet an dunder attribute `__gt_dims__` that has then that functionality.

jdahm commented on src/gt4py/storage/storage.py:
```diff
@@ -39,659 +36,105 @@ def _error_on_invalid_backend(backend):
 
 
-def empty(backend, default_origin, shape, dtype, mask=None, *, managed_memory=False):
+def empty(backend, aligned_index, shape, dtype, dimensions=None):
```
Can `aligned_index` be made optional and default to `(0, 0, 0)`?

gronerl replied:
Docstrings done.

jdahm replied:
Sorry to bother, but I still don't see docstrings on `empty`, `ones`, and `zeros` here. Can you add those before merging? Thanks!

gronerl replied:
They're there! I think if you click the file name above it takes you to the old version.

jdahm commented:
It seems somehow inconsistent to allow (and currently require) aligning memory on an index, but then not defaulting to that as the origin. I understand the reason behind it is that we want "no storages", so we can't rely on attributes being present. In order to have maximum performance then the user has to specify an `aligned_index` in one part of the code, and then use _exactly_ that value everywhere else it's called. Seems like duplication, but right now I don't see a great way around it.

jdahm commented on docs/gt4py/quickstart.rst:
```diff
@@ -183,29 +185,43 @@ set of routines used in `NumPy` for allocation: ``from_array``, ``ones``, ``zero
 
-#. If when calling the stencil, no other `origin` is specified, this value is where the `iteration domain` begins, i.e.
-   the grid point with the lowest index where a value is written.
+When passing buffers to stencils, they can be in any form that is compatible with ``np.asarray`` or ``cp.asarray``,
```
Is the idea that it's really anything that implements the array interface protocol (buffer protocol) https://numpy.org/doc/stable/reference/arrays.interface.html?

gronerl replied:
In principle, yes. e.g. `numpy` backends and `dace:*` backends should be fine with that. The option to reject a buffer based on wrong layout, (including no contiguous dimension) is still there!

jdahm replied:
Great!

jdahm commented on src/gt4py/backend/dace_backend.py:
```diff
@@ -366,13 +365,17 @@ def unique_index(self) -> int:
+                res[
+                    name
+                ] = "py::{pybind_type} {name}, std::array<gt::int_t,{ndim}> {name}_origin".format(
+                    pybind_type="object"
+                    if self.backend.storage_info["device"] == "gpu"
+                    else "buffer",
```
Just curious: why do these need to be different? I'm sure there's a good reason. I'm not too familiar with the details of pybind11.

gronerl replied:
They should have not been `py::buffer` even now. It works now, since the GPU storages are both a `py::buffer` (due to the dual CPU buffer)  but also have the `__cuda_array_interface__` attribute that `as_cuda_sid` accesses. However, now we pass `CuPy` arrays from the python side, so the old interface breaks down.

jdahm reviewed (commented):
Looks good. I wrote down a few thoughts to consider.

gronerl commented:
The failing tests should pass after merging/squashing/rebasing master. As overriding gitlab is currently disable, can @jdahm  please do so and then `bors try` again?

gronerl commented on docs/gt4py/arrays.rst:
```diff
@@ -0,0 +1,175 @@
+For the performance-optimal allocation and initialization of arrays to be used in GT4Py, we provide the following set of
+functions which closely resemble their NumPy counterparts (meaning of the common parameters is explained below).
+
+The return type is either a ``numpy.ndarray`` or a `cupy.ndarray`, for CPU and GPU backends, respectively.
```
No, I think this did the roundtrip to markdown and back once too many, I'll make a pass through it to add the `:code:` consistently.

gronerl commented on src/gt4py/storage/utils.py:
```diff
@@ -12,72 +12,84 @@
+            else list("IJK") + [str(d) for d in range(len(shape) - 3)]
+        )
 
+    if not all(
+        isinstance(d, (str, Axis)) and (str(d).isdigit() or str(d) in "IJK") for d in dimensions
+    ):
+        raise ValueError(f"Invalid dimensions definition: '{dimensions}'")
```
I think this should be a user-facing error message, not a debug-stile assertion / internal consistency assumption. Why do you think an assertion is more suitable?

jdahm replied:
I believe I was thinking about performance at some point, but this isn't called in the `StencilObject` after all, only in the allocation methods. In that case it's correct to have this be a `ValueError`.

gronerl commented on tests/test_integration/test_stencil_object.py:
```diff
@@ -34,11 +36,13 @@ def stencil(in_field: Field[float], out_field: Field[float], *, offset: float):
     shape = (4, 4, 4)
-    in_storage = gt_storage.ones(
-        backend=backend, default_origin=(0, 0, 0), shape=shape, dtype=float
+    in_storage = OriginWrapper(
+        array=gt_storage.ones(backend=backend, aligned_index=(0, 0, 0), shape=shape, dtype=float),
+        origin=(0, 0, 0),
```
To be clear, `OriginWrapper` etc are just utility wrappers for testing. There won't be anything user-faceing like that. I think it isn't worth to have additional API just to avoid that small repetition.

gronerl commented:
Switching this to draft while I'm working on addressing review comments.

gronerl commented:
bors try

gronerl commented:
bors try-
bors try

gronerl commented on src/gt4py/stencil_object.py:
```diff
@@ -17,34 +17,76 @@
+def _compute_cache_key(array_infos, parameter_args, domain, origin) -> int:
     field_data = tuple(
-        (name, field.shape, getattr(field, "default_origin", (0, 0, 0)))
-        for name, field in field_args.items()
-        if field is not None
+        (name, info.array.shape, info.origin or (0, 0, 0))
+        for name, info in array_infos.items()
+        if info is not None
     )
     return hash((field_data, *parameter_args.keys(), dumps(domain), dumps(origin)))
```
I changed it so we need `asarray(...).shape` and `getattr(arg, "__gt_dims__", (0, 0, 0))` I.e. there is an additional asarray call compared to before. I would assume that's still efficient enough.

jdahm replied:
Thanks. That's great!

gronerl commented on docs/gt4py/arrays.rst:
```diff
@@ -0,0 +1,175 @@
+
+The return type is either a ``numpy.ndarray`` or a `cupy.ndarray`, for CPU and GPU backends, respectively.
+
+:code:`empty(shape: Sequence[int], dtype: dtype_like = np.float64, **kwargs) -> ndarray`
+    Allocate an array with uninitialized (undefined) values.
```
Fixed.

gronerl commented on tests/test_integration/test_dace_parsing.py:
```diff
@@ -81,20 +83,25 @@ def defn(outp: gtscript.Field[np.float64], par: np.float64):
-    outp = gt_storage.zeros(
-        dtype=np.float64, shape=(10, 10, 10), default_origin=(0, 0, 0), backend=backend
+    # DaCe in the currently tagged version of this repo (~=0.13 doesn't support non-numpy arrays)
+    wrapper_type = OriginWrapper if "gpu" in backend else NdarraySubclassOriginWrapper
```
dace released 0.14 by now where this is no longer needed. I updated the setup.cfg to use it.

gronerl commented on tests/test_unittest/test_storage.py:
```diff
@@ -445,619 +388,59 @@ def test_cpu_constructor(alloc_fun, backend):
-    mask = param_dict["mask"]
-    default_origin = param_dict["default_origin"]
+
+    import cupy as cp
```
Probably a mistake.

gronerl replied:
fixed.

gronerl commented on src/gt4py/stencil_object.py:
```diff
@@ -17,34 +17,76 @@
+    dimensions: Optional[Tuple[str]] = None
+
+
+def _extract_array_infos(field_args, device) -> Dict[str, Optional[_ArgsInfo]]:
```
Done.

gronerl commented:
I addressed all comments where I agree and marked as resolved where I deemed it appropriate. We need to discuss making `aligned_index` optional. Please mark other conversations as resolved if your issue is addressed. Or leave further comments.

gronerl commented:
bors try

jdahm commented on docs/gt4py/quickstart.rst:
```diff
@@ -183,29 +185,43 @@ set of routines used in `NumPy` for allocation: ``from_array``, ``ones``, ``zero
     field_b = gt_storage.ones(
-        backend=backend, shape=(10, 10, 10), dtype=np.float64, default_origin=(0, 0, 0)
+        backend=backend, shape=(10, 10, 10), dtype=np.float64, aligned_index=(0, 0, 0)
     )
     field_c = gt_storage.zeros(
-        backend=backend, shape=(10, 10, 10), dtype=np.float64, default_origin=(0, 0, 0)
+        backend=backend, shape=(10, 10, 10), dtype=np.float64, aligned_index=(0, 0, 0)
     )
     result = gt_storage.empty(
-        backend=backend, shape=(10, 10, 10), dtype=np.float64, default_origin=(0, 0, 0)
+        backend=backend, shape=(10, 10, 10), dtype=np.float64, aligned_index=(0, 0, 0)
     )
```
Or perhaps reformat these to
```python
result = gt_storage.empty(
    backend, (0, 0, 0), shape=(10, 10, 10), dtype=np.float64
)
```

as the function signature suggests.

gronerl replied:
I reformatted according to the signature.

jdahm commented on tests/test_integration/test_code_generation.py:
```diff
@@ -527,26 +544,23 @@ def k_to_ijk(outp: Field[np.float64], inp: Field[gtscript.K, np.float64]):
+        aligned_index=(0,),
         dtype=np.float64,
-        mask=[False, False, True],
+        dimensions="K",
```
Should this be `dimensions=["K"],`?

gronerl replied:
A string is also a sequence of strings, which means either is fine, (unless there are >10 data dimensions).

jdahm reviewed (approved):
Great work on this!

egparedes commented on docs/gt4py/arrays.rst:
```diff
@@ -0,0 +1,175 @@
+
+GT4Py provides utilities to allocate buffers that have optimal layout and alignment for a given backend.
+
+In this document, we describe the interfaces for
+* supported buffer interfaces
```
```suggestion
In this document, we describe the interfaces for

* supported buffer interfaces
```

egparedes commented on docs/gt4py/arrays.rst:
```diff
@@ -0,0 +1,175 @@
+Dimension Mapping
+^^^^^^^^^^^^^^^^^
+
+The user can optionally implement a :code:`__gt_dims__` attribute in the object implementing any of the supported buffer
+interfaces. As a fallback if neither is specified the dimensions given in the annotations (by means of
+:code:`gtscript.Field`) are assumed. The returned object should be a tuple of strings labeling the dimensions in index
+order.
```
```suggestion
The user can optionally implement a :code:`__gt_dims__` attribute in the object implementing any of the supported buffer
interfaces. The returned object should be a tuple of strings labeling the dimensions in index order.
As a fallback if the attribute is not implemented, it is assumed that the buffer contains the dimensions given in the annotations
(by means of :code:`gtscript.Field`) exactly in the same order.
```

egparedes commented on docs/gt4py/arrays.rst:
```diff
@@ -0,0 +1,175 @@
+    For common keyword-only arguments, please see below.
+
+:code:`from_array(data: array_like, device_data: array_like = None, *, dtype: dtype_like = np.float64, **kwargs) -> ndarray`
+    Used to allocate an array with values initialized to those of a given array.
```
```suggestion
    Used to allocate an array with values initialized from the content of a given array.
```

egparedes commented on docs/gt4py/arrays.rst:
```diff
@@ -0,0 +1,175 @@
+        + :code:`device_data: array_like`. The original array in case copying to a gpu buffer is
+          desired. The same buffer could also be passed through :code:`data` in that case, however this
+          parameter is here to provide the same interface like the :code:`as_storage` function.
+
+        + :code:`sync_state: gt4py.storage.SyncState`. If `managed="gt4py"` indicates which of the
+          provided buffers, :code:`data` or :code:`device_data`, is up to date at the time of initialization.
```
```suggestion
```

egparedes commented on docs/gt4py/arrays.rst:
```diff
@@ -0,0 +1,175 @@
+    Parameters:
+        + :code:`data: array_like`. The original array from which the storage is initialized.
+
+        + :code:`device_data: array_like`. The original array in case copying to a gpu buffer is
+          desired. The same buffer could also be passed through :code:`data` in that case, however this
+          parameter is here to provide the same interface like the :code:`as_storage` function.
```
Is `device_data` still needed/relevant? Cannot be removed?

egparedes reviewed (commented):
Just a partial review for now (only the docs part).

egparedes commented on src/gt4py/storage/storage.py:
```diff
@@ -39,659 +44,372 @@ def _error_on_invalid_backend(backend):
+        aligned_index=aligned_index,
+        dimensions=dimensions,
+    )
+    storage[...] = fill_value
```
Same suggestion as above about doing a explicit type casting of the fill value

egparedes commented on src/gt4py/storage/storage.py:
```diff
@@ -39,659 +44,372 @@ def _error_on_invalid_backend(backend):
+        aligned_index=aligned_index,
+        dimensions=dimensions,
     )
     storage[...] = 0
```
Same comment as as above

egparedes commented on src/gt4py/storage/storage.py:
```diff
@@ -39,659 +44,372 @@ def _error_on_invalid_backend(backend):
     if dtype is None:
-        dtype = xp.asarray(data).dtype
+        dtype = asarray(data).dtype
+    dtype = np.dtype(dtype)
```
If `cupy.ndarray.dtype` returns `np.dtype` instances, then I guess this is redundant... although most likely a bit safer
```suggestion
```

gronerl replied:
dtype is only derived from `data` if it is `None` in the explicit argument. Otherwise, it can be e.g. `str` instead and needs to be marshalled through `np.dtype`

egparedes replied:
Sure, you're fully right here. My mistake.

egparedes commented on src/gt4py/storage/utils.py:
```diff
@@ -12,72 +12,85 @@
+
+
+class CudaArrayInterfaceType(Protocol):
+    __array_interface__: Dict[str, Any]
```
```suggestion
    __cuda_array_interface__: Dict[str, Any]
```

egparedes commented on src/gt4py/storage/utils.py:
```diff
@@ -217,44 +215,70 @@ def allocate_gpu_unmanaged(default_origin, shape, layout_map, dtype, alignment_b
-        return raw_buffer, raw_buffer
+def allocate_cpu(aligned_index, shape, layout_map, dtype, alignment_bytes):
+    dtype = np.dtype(dtype)
+    assert (
+        alignment_bytes % dtype.itemsize
+    ) == 0, "Alignment must be a multiple of byte-width of dtype."
```
Unless this check uses `assert` on purpose for performance reasons, I think it'd be better to use `if` and raise an exception here to avoid it can be disabled in performance mode.

egparedes reviewed (commented):
Second part of my review.

gronerl commented on docs/gt4py/quickstart.rst:
```diff
@@ -180,32 +182,46 @@ set of routines used in `NumPy` for allocation: ``from_array``, ``ones``, ``zero
-        default_origin=(0, 0, 0),
+        aligned_index=(0, 0, 0),
     )
     field_b = gt_storage.ones(
```
Do we need one? It's a quick start, not a comprehensive documentation. Should be straight-forward.

egparedes replied:
Fair enough.

gronerl commented on src/gt4py/storage/storage.py:
```diff
@@ -28,8 +30,11 @@
+if np.lib.NumpyVersion(np.__version__) >= "1.20.0":
+    from numpy.typing import ArrayLike, DTypeLike
+else:
+    ArrayLike = Any
+    DTypeLike = Any
```
If we're going to approximate, I'd rather approximate "from above" i.e. include more in the type then is allowed. (e.g. your diff does not yet include `__array_interface__`, `__cuda_array_interface__` and others. I think `Any` is just good enough given that it's only a fallback for old NumPy versions. 

It's a lot of clutter that leads to type checkers falsly flagging correct code.

egparedes replied:
Fair enough.

gronerl commented:
I fixed as suggested or left a comment on all your comments. @egparedes

egparedes reviewed (approved):
It looks good.

gronerl commented:
bors try

egparedes commented:
bors try

gronerl commented:
bors try

egparedes commented:
bors try
